### ISSUE TITLE
Write bundle to disk

### ISSUE THREAD

**Issue by KyleAMathews** [webpack/webpack-dev-server#62]
I'm developing an app that has several entry points — one of which needs a custom html file that I'm generating using a View in my hapi.js server.

Is there a way I can configure the dev-server to write the generated js files to disk? Right now I either have to have another webpack watcher to compile files or stop and manually compile the files when I make a change to that particular part of the code.

**Comment by max-mykhailenko** [comment]
+1

**Comment by sokra** [comment]
Writing to disk is not an option for the webpack-dev-server. Instead you can use `webpack --watch` which does the same but writes files to disk + a static fileserver.

**Comment by diurnalist** [comment]
@sompylasar this is a very late reply, but you can examine the generated assets if you visit your dev server under the /webpack-dev-server path. I.e. if you're running it on 9000, http://localhost:9000/webpack-dev-server.

**Comment by sompylasar** [comment]
@diurnalist Sure, I know. This is how I do debugging in the browser. But I wanted to have the files in my text editor rather than in the browser's developer tools.

**Comment by KyleAMathews** [comment]
@sokra but that means you can't do hot-reloads...

**Comment by diurnalist** [comment]
@KyleAMathews how so? Nothing about hot reloads requires the assets be written to disk. I'm using it in a project right now without issue.

**Comment by KyleAMathews** [comment]
@diurnalist sorry wasn't clear — I want assets written to the disk. I was the one who originally created this issue ;-)

Right now I have to run both `webpack-dev-server` and `webpack -w`.

**Comment by diurnalist** [comment]
@KyleAMathews hmm, I'm still trying to understand why you need to run two webpack compilations. It seems irrelevant how the bundle is served. The only thing I can think is that somehow you need to do additional processing outside of webpack land on one of your output files? Feel free to send me an email (listed on my profile) - I've had to deal with a few weird things when integrating webpack into a big project, maybe my experience can be of use.

**Comment by max-mykhailenko** [comment]
@KyleAMathews you can try setup git hook and run webpack on commit, pull etc.

**Comment by KyleAMathews** [comment]
@diurnalist I have two entry points, one of which I want served by webpack-dev-server for hot reloading (an admin page) and the other entry point I need compiled to disk so it can be loaded as part of a server generated React page.

@max-mykhailenko I need to see the results of my changes before committing ;-)

**Comment by sokra** [comment]
> I have two entry points

You should have two configurations, because the `target` option differs: `web` and `node`.

**Comment by KyleAMathews** [comment]
@sokra I was a bit unclear — the target for both entry points is the web — React is used to generate the initial html page that's sent but I still need the compiled js to animate the react app in the browser.

**Comment by diurnalist** [comment]
@KyleAMathews so, just to be clear (again, happy to discuss via email, these GH comment reply flows get a bit ridic after a certain point...) - does your generated HTML file need to include the script inlined? Or is it fine to include it as an external link? Obviously if the latter is true, you can simply link your script in:

```
<script src="http://localhost:9090/js/your-react-bundle.js"></script>
```

You'd just need to be able to change the host for when you deploy the app to a remote server, obviously.

If you need to actually inline your script... that's a different issue. But, I'd ask: why would you need to do that? For what it's worth, it's possible. I'm currently doing it to bootstrap some page-specific JSON into my app when it starts. The tricky part here is you need to understand what IDs webpack has assigned your modules. To do that, you need to parse the stats object webpack returns when a compilation completes. I have some code I could open-source that helps with this.

Basically, I have some script template that looks something like this:

```js
webpackJsonp([], {
  0: function (module, exports, require) {
    var App = require(__APPLICATION_MODULE_ID__);
    App.start(__BOOTSTRAP_DATA__);
  }
});
```

And then when I'm generating my HTML page, I just read that template and replace the tokens with the webpack module id for the App module and the JSON string of bootstrap data. This works essentially by injecting a dummy webpack chunk on to the page that has no dependencies. You would need to ensure that you have loaded the main webpack library on the page before doing this, or else the `webpackJsonp` function will not exist yet.

**Comment by QuantumInformation** [comment]
I ran `webpack-dev-server webpack -w`in windows and get this:

```
WARNING in ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/README.md
Module parse failed: C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_modules\chokidar\node_modules\readdirp\no
de_modules\minimatch\README.md Line 1: Unexpected token ILLEGAL
You may need an appropriate loader to handle this file type.
| # minimatch
| 
| A minimal matching utility.
 @ ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch ^\.\/.*$

ERROR in ./~/webpack/lib/node/NodeMainTemplatePlugin.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\lib\node
 @ ./~/webpack/lib/node/NodeMainTemplatePlugin.js 157:4-17 172:4-17

ERROR in ./~/webpack/package.json
Module parse failed: C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\package.json Line 2: Unexpected token :
You may need an appropriate loader to handle this file type.
| {
|   "name": "webpack",
|   "version": "1.8.4",
|   "author": {
 @ ./~/webpack/lib/Stats.js 104:16-42

ERROR in ./~/webpack/~/uglify-js/tools/node.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\uglify-js\tools
 @ ./~/webpack/~/uglify-js/tools/node.js 2:9-22

ERROR in ./~/webpack/lib/node/NodeOutputFileSystem.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\lib\node
 @ ./~/webpack/lib/node/NodeOutputFileSystem.js 5:9-22

ERROR in ./~/webpack/lib/node/OldNodeWatchFileSystem.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\lib\node
 @ ./~/webpack/lib/node/OldNodeWatchFileSystem.js 5:9-22

ERROR in ./~/webpack/~/mkdirp/index.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\mkdirp
 @ ./~/webpack/~/mkdirp/index.js 2:9-22

ERROR in ./~/webpack/~/enhanced-resolve/~/graceful-fs/fs.js
Module not found: Error: Cannot resolve module 'module' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\enhanced-resolve\n
ode_modules\graceful-fs
 @ ./~/webpack/~/enhanced-resolve/~/graceful-fs/fs.js 5:10-27

ERROR in (webpack)/~/node-libs-browser/~/constants-browserify/constants.json
Module parse failed: C:\Users\Laptop\AppData\Roaming\npm\node_modules\webpack-dev-server\node_modules\webpack\node_modules\node-libs-browser\node_modules\constants-browserif
y\constants.json Line 2: Unexpected token :
You may need an appropriate loader to handle this file type.
| {
|   "O_RDONLY": 0,
|   "O_WRONLY": 1,
|   "O_RDWR": 2,
 @ ./~/webpack/~/enhanced-resolve/~/graceful-fs/polyfills.js 2:16-36

ERROR in ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/max-open.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\enhanced-resolve\node
_modules\graceful-fs\test
 @ ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/max-open.js 1:11-25

ERROR in ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/readdir-sort.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\enhanced-resolve\node
_modules\graceful-fs\test
 @ ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/readdir-sort.js 1:11-25

ERROR in ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/open.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\enhanced-resolve\node
_modules\graceful-fs\test
 @ ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/open.js 1:11-25

ERROR in ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/write-then-read.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\enhanced-resolve\node
_modules\graceful-fs\test
 @ ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/write-then-read.js 4:11-25

ERROR in ./~/webpack/~/watchpack/~/chokidar/index.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_modules
\chokidar
 @ ./~/webpack/~/watchpack/~/chokidar/index.js 3:9-22

ERROR in ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/write-then-read.js
Module not found: Error: Cannot resolve module 'rimraf' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\enhanced-resolve\n
ode_modules\graceful-fs\test
 @ ./~/webpack/~/enhanced-resolve/~/graceful-fs/test/write-then-read.js 2:13-30

ERROR in ./~/webpack/~/watchpack/~/graceful-fs/fs.js
Module not found: Error: Cannot resolve module 'module' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_mod
ules\graceful-fs
 @ ./~/webpack/~/watchpack/~/graceful-fs/fs.js 5:10-27

ERROR in ./~/webpack/~/watchpack/~/chokidar/lib/nodefs-handler.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_modules
\chokidar\lib
 @ ./~/webpack/~/watchpack/~/chokidar/lib/nodefs-handler.js 3:9-22

ERROR in ./~/webpack/~/watchpack/~/chokidar/lib/fsevents-handler.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_modules
\chokidar\lib
 @ ./~/webpack/~/watchpack/~/chokidar/lib/fsevents-handler.js 3:9-22

ERROR in ./~/webpack/~/watchpack/~/graceful-fs/test/readdir-sort.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\graceful-fs\test
 @ ./~/webpack/~/watchpack/~/graceful-fs/test/readdir-sort.js 1:11-25

ERROR in ./~/webpack/~/watchpack/~/graceful-fs/test/max-open.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\graceful-fs\test
 @ ./~/webpack/~/watchpack/~/graceful-fs/test/max-open.js 1:11-25

ERROR in ./~/webpack/~/watchpack/~/graceful-fs/test/open.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\graceful-fs\test
 @ ./~/webpack/~/watchpack/~/graceful-fs/test/open.js 1:11-25

ERROR in ./~/webpack/~/watchpack/~/graceful-fs/test/write-then-read.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\graceful-fs\test
 @ ./~/webpack/~/watchpack/~/graceful-fs/test/write-then-read.js 4:11-25

ERROR in ./~/webpack/~/watchpack/~/graceful-fs/test/write-then-read.js
Module not found: Error: Cannot resolve module 'rimraf' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_mod
ules\graceful-fs\test
 @ ./~/webpack/~/watchpack/~/graceful-fs/test/write-then-read.js 2:13-30

ERROR in ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/graceful-fs/graceful-fs.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_modules
\chokidar\node_modules\readdirp\node_modules\graceful-fs
 @ ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/graceful-fs/graceful-fs.js 3:26-39

ERROR in ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/graceful-fs/polyfills.js
Module not found: Error: Cannot resolve module 'fs' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_modules
\chokidar\node_modules\readdirp\node_modules\graceful-fs
 @ ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/graceful-fs/polyfills.js 1:9-22

ERROR in ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/basic.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\chokidar\node_modules\readdirp\node_modules\minimatch\test
 @ ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/basic.js 6:10-24

ERROR in ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/caching.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\chokidar\node_modules\readdirp\node_modules\minimatch\test
 @ ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/caching.js 2:10-24

ERROR in ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/brace-expand.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\chokidar\node_modules\readdirp\node_modules\minimatch\test
 @ ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/brace-expand.js 1:10-24

ERROR in ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/extglob-ending-with-state-char.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\chokidar\node_modules\readdirp\node_modules\minimatch\test
 @ ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/extglob-ending-with-state-char.js 1:11-25

ERROR in ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/defaults.js
Module not found: Error: Cannot resolve module 'tap' in C:\Users\Laptop\WebstormProjects\angular-webpack-lazy-loading\node_modules\webpack\node_modules\watchpack\node_module
s\chokidar\node_modules\readdirp\node_modules\minimatch\test
 @ ./~/webpack/~/watchpack/~/chokidar/~/readdirp/~/minimatch/test/defaults.js 6:10-24

```

**Comment by QuantumInformation** [comment]
I basically want the server auto refreshing when I make changes in the code, similar to how the ember-cli works.

**Comment by wonnage** [comment]
This would be a really nice thing to have... I'm writing some isomorphic JS; it currently requires a `webpack --watch` to generate a bundle.js that the server can use, and a `webpack-dev-server` to serve a bundle.js for the browser.

**Comment by NickPresta** [comment]
I've tried to follow allow this thread and the linked threads, but I'm still unsure of my current options.

It is possible to run Webpack Dev Server in "hot reload" mode (to support React Hot Loader, etc) while still writing the files to disk?

**Comment by JorritPosthuma** [comment]
I'd like to add an argument for adding this option 😊. When using electron, in certain cases, you are only allowed to use files from a `file:` path: https://github.com/atom/electron/blob/master/atom/renderer/lib/web-view/web-view-attributes.coffee#L201

@nickpresta As far as I know, not right now. You have to run a separate `webpack -w`.

**Comment by thelinuxlich** [comment]
+1 for dev server writing to disk!

**Comment by sheerun** [comment]
We also needed it for node's native modules (`.node`). They can't be loaded from web server.

**Comment by mrberggg** [comment]
+1 for writing to disk. Just adds confusion when starting out.

**Comment by ehgoodenough** [comment]
+1 for not only serving the compiled files, but also writing the compiled files to disk.

**Comment by gajus** [comment]
This would be useful when using webpack + BrowserSync to refresh when static assets (ie., CSS) change.

**Comment by gajus** [comment]
I have released a plugin that forces webpack-dev-server to write to the file system.

https://github.com/gajus/write-file-webpack-plugin

**Comment by sompylasar** [comment]
@gajus Thank you for that!

**Comment by Sridatta19** [comment]
@gajus  thank you.

+1 Still hoping for an official resolution to be provided

**Comment by batmandarkside** [comment]
+1

**Comment by wailorman** [comment]
+1 Make it optional at least, please

**Comment by ehgoodenough** [comment]
A good solution is to just use Webpack, which actually writes to the disk, which you can pair with something like BrowserSync. It seems like WebpackDevServer is meant only for non-writing-to-disc development. :]

**Comment by gajus** [comment]
There is a plugin for this. Whats everyone upset about? Just use it.

**Comment by ehgoodenough** [comment]
@gajus Drop a link, bro! :]

**Comment by gajus** [comment]
https://github.com/webpack/webpack-dev-server/issues/62#issuecomment-151343141

**Comment by wailorman** [comment]
@gajus 
Unfortunately, your plugin writes only bundle.js and skips other files (e.g. images, fonts). Although its not crucial. Using data-urls for this stuff will solve problem.

I was having a lot of problems with launching web-dev-server via gulp and writing right config for it. So, I started using ```webpack-dev-server --content-base dist/``` and all problems went out! Highly recommend to use this way.

**Comment by arcseldon** [comment]
+1 for a flag that permits writing contents out to disk.

**Comment by gajus** [comment]
> Unfortunately, your plugin writes only bundle.js and skips other files (e.g. images, fonts). Although its not crucial. Using data-urls for this stuff will solve problem.

It writes the same what `webpack` program would. If you have not configured webpack program to write whatever additional resources, it will not write them.

**Comment by gajus** [comment]
> Unfortunately, your plugin writes only bundle.js and skips other files (e.g. images, fonts). Although its not crucial. Using data-urls for this stuff will solve problem.

@wailorman This has been fixed in the latest release. https://github.com/gajus/write-file-webpack-plugin/issues/14

**Comment by wailorman** [comment]
@gajus Thank you very much! Your plugin is really awesome

**Comment by richburdon** [comment]
@sokra 
> But why would you need to access the generated assets?

Suppose you're writing a Chrome Extension. These aren't loaded from a server. The bundle file is needed to be loaded from disk. Not adding this option requires setting up a separate Grunt watch task just to call webpack. That's messy since for other bundles I do want to load from the "dev server".

**Comment by ahmadster** [comment]
@gajus perfect! exactly what I needed. Thx!

**Comment by justinjdickow** [comment]
I'll add my super random need for this as a +1.

I am working on a proof of concept for a product that has an HTML5 UI on an embedded (vehicle) OS. The HTML talks to the embedded component over WebSocket and does not have any sort of internet access. There are points in the communication which the embedded component passes a path to the HTML5 app containing assets such as images for display. I am currently creating a tunnel so the app can grab the static files via ip + file path and running nginx on the embedded component to serve the static files (hack) while I run the web app from my mac on the same network. It would be ideal if I could run these on the same OS and launch the index.html file (referencing the compiled index.js file) into the web view so that its url is file:// instead of http:// so that it has access to the file system. To do this I need the compiled index.js (I think)

**Comment by gajus** [comment]
> @justinjdickow 
> I'll add my super random need for this as a +1.

https://github.com/gajus/write-file-webpack-plugin

**Comment by justinjdickow** [comment]
@gajus I did end up finding that and it worked perfectly for my needs. Thank you!

**Comment by anuja8275** [comment]
@diurnalist ,@KyleAMathews can i have any single command which will able to create my bundle.js file in my app not with --watch and that should able to run webpack-dev-server also

**Comment by slorber** [comment]
Another usecase:

- I have a Cordova app based on my mobile website
- I want to dev my mobile website with hot reloading in the browser
- To test my app on real device, all assets must be available in /cordova/www folder

Currently, whenever I'm working on the mobile app, I have to fallback to `webpack --watch`, and doing so means I don't have anymore hot reloading when "previewing" my devs in the browser. I need both hot reloading + writing to disk

Will try plugin of @gajus thanks

**Comment by Hosar** [comment]
@gajus excelent plugin, thanks!

**Comment by Kos** [comment]
Thank you for the plugin @gajus! For the record, my use case is I'm running selenium tests which expect a bundle to be there on the file system. Currently we're all running webpack dev server, but whenever someone wants to run a selenium test they need to remember to kill the dev server and run webpack --watch instead.

**Comment by chikara-chan** [comment]
My solution is putting "emit" hook into the build process.
```js
const compiler = webpack(config)
compiler.plugin('emit', (compilation, callback) => {
    const assets = compilation.assets
    let file, data
    Object.keys(assets).forEach(key => {
        // if (key.match(/\.html$/)) {
            file = path.resolve(__dirname, key)
            data = assets[key].source()
            fs.writeFileSync(file, data)
        // }
    })
    callback()
})
app.use(devMiddleware(compiler, {
    noInfo: true,
    publicPath: config.output.publicPath
}))
app.use(hotMiddleware(compiler))

**Comment by georgiosd** [comment]
Very strange that this is not supported. I'm using webpack to write an electron app and since the main module can't really be hot-reloaded, this is a pain in the behind. I have to stop the dev server every time, build, and watch again.

**Comment by gajus** [comment]
> I have to stop the dev server every time, build, and watch again.

Just use https://github.com/gajus/write-file-webpack-plugin

**Comment by slorber** [comment]
yes @gajus that works fine for me, not sure it's worth integrating in webpack core

(btw not sure it's very useful to write hot updates to disk by default)

**Comment by simon-paris** [comment]
It doesn't seem right that I need a plugin for this.

Our use case is this: Some of our developers prefer `webpack --watch` and some prefer `webpack-dev-server`. We have a button in our app to switch between them. But because the dev server doesn't output files, I need to build with plain webpack first so that I can reach that button.

**Comment by gajus** [comment]
> It doesn't seem right that I need a plugin for this.

Whats the problem with using a plugin?

**Comment by simon-paris** [comment]
> Whats the problem with using a plugin?

I feel like this should be a built in behavior.

**Comment by gajus** [comment]
The core features are dictated by the mainstream use. This isn't a mainstream use case. Plugins are to support alternative use cases.

**Comment by ashish-chopra** [comment]
Thanks @gajus. After a long search related to emit usecase i landed upto this thread and found your plugin a Savior. THanks. +1 for your plugin.

**Comment by cchamberlain** [comment]
I am writing a C# => TypeScript transpiler plugin. TypeScript editors / type checkers require hard files on disk. Currently detecting if I'm running in watch and using `fs.writeFile` to emit assets, otherwise I use the assets object. It would be nice to be able to route assets to avoid less differences in dev / prod code paths.

**Comment by SteveGBanton** [comment]
@anuja8275 @simon-paris @ashish-chopra What if you just use: 

"start": "./node_modules/.bin/webpack-dev-server && ./node_modules/.bin/webpack"

in your package.json? Dev server runs first and you make your edits. When done you ctrl-c to close the terminal process and webpack will automatically start to create the file system bundle.

**Comment by SteveGBanton** [comment]
,

**Comment by Kos** [comment]
@SteveGBanton I don't think that would provide good usability - most users expect Ctrl+C to be instantenous and a webpack run can easily take half a minute.

**Comment by feusebio** [comment]
I had some problems to get file bundled in dir that i want, the problem was because i run webpack in Windows, and i had to change the slacks in paths.
`
 Ex: context: __dirname + '/resources/assets/js', to context: __dirname + '\\resources\\assets\\js', 
`

**Comment by SteveGBanton** [comment]
@Kos You're right, might be a bit awkward in some cases. This also works for me:

"scripts": {
    "start": "./node_modules/.bin/webpack-dev-server",
    "start-disk": "./node_modules/.bin/webpack --watch"
  }

Run 'npm start' first, then 'npm run start-disk' in a separate terminal window. You can develop with webpack-dev-server and bundle is automatically written to disk at the same time through webpack --watch... Might need to delete your bundle file to start.

**Comment by asumaran** [comment]
I've spent 20 minutes reloading and deleting `node_modules` and reinstalling everything just to find out `webpack-dev-server` doesn't write to disk. Usually I want to check what in the bundle is changing or sometimes I want to check how the code is generated (for technical purposes) That's why I think it's useful to have an option to write the generated files to disk so I can check them in my own editor to inspect changes.

**Comment by herzinger** [comment]
@gajus great plugin, thanks!

That being said, I'm also of the opinion it should be integrated in the core project, as an optional flag. I don't even have a complicated user case, I just like to see what's being outputted right there in my ide whenever I want to. I don't believe I'm alone in that.

**Comment by atrauzzi** [comment]
Working on an electron app.  Would like to be able to serve the initial main from the filesystem, but have subsequent loads work via the server.

**Comment by veetil09** [comment]
@gajus Hey, thanks for the plugin. You may want to surround the ```fs.writeFileSync(relativeOutputPath.split('?')[0], assetSource);``` in a try-catch since it can throw an exception.

**Comment by gajus** [comment]
> @gajus Hey, thanks for the plugin. You may want to surround the fs.writeFileSync(relativeOutputPath.split('?')[0], assetSource); in a try-catch since it can throw an exception.

PR welcome.

**Comment by veetil09** [comment]
https://github.com/gajus/write-file-webpack-plugin/pull/47

**Comment by valorloff** [comment]
This feature is highly needed for HMR universal apps, when need re-bundle server and send changed client.bundle to the client

**Comment by heisian** [comment]
@everyone I see at least two viable solutions provided here in this thread so let's make a couple things clear before we keep asking to add more stuff:
1) ANY addition to webpack functionality is by nature a plugin. So if you think writing to disk should be an option and not a plugin, guess what, any way you slice it it will be a (internal) plugin.
2) @gajus 's solution which I'm about to try out! https://github.com/gajus/write-file-webpack-plugin
3) @chikara-chan 's super easy plugin which I will try first.

Now, the reason **I** want write-to-disk is b/c I need to be able to make server-side changes without having to re-compile the client from scratch when the server restarts. I've separated out the processes, but the only way I see the forked server process being able to get the references to the webpack assets in-memory is via node's IPC channels (someone correct me if I'm wrong!). While probably the best solution, I don't really want to mess with it when serving static files on disk is trivial in comparison. Being able to use HMR w/ that absolute path on -disk would be an easy path. Let's see if this works...

**Comment by heisian** [comment]
Alright the above doesn't really work because `webpack-hot-client` does us the gracious service of updating browser assets via websocket and I was trying to cut that part out completely.

So still need a mechanism to have the server server and the client server as separate processes... more easily done by simply having two servers on two different ports.

**Comment by paulomunoz** [comment]
@sokra:
> But why would you need to access the generated assets? This shouldn't be required.

I would normally agree with you, but at this moment I am trying to use Visual Studio Code debugging tools to debug my app, and it only works if you provide it with the file location in disk. I used https://github.com/gajus/write-file-webpack-plugin to do the trick, and that worked fine.

Normally I would not want it to be written in the disk though, so I'll see if I can separate this in a different npm script or something like that!

**Comment by Matchlighter** [comment]
Just my two cents: My project's backend is written in Python using Django. The Webpack HTTP Plugin just wouldn't cut it for my multiple-entry application, so I wrote a plugin to make Webpack write a file with all of the chunks it output and all of their required chunks. Django could then read that file and add the appropriate HTML tags. That blows up when Webpack doesn't spit out any files.
I will admit that the use case are few, but they do exist. @gajus plugin worked dandily. At least an official link in the README would be nice.

**Comment by TacB0sS** [comment]
Extremely late for the party, but my 2 cents on this:

You are correct that when you are busy in developing your page and components and styles, this is somewhat redundant, but when you are configuring your environment, adding third party sdks, scripts, configurations, files (and god know what other requirements they have) all that while maintaining dev and prod (maybe also staging) configurations.. things gets complicated, and although I can just execute webpack instead of webpack-dev-server and see the results, I do prefer to catch as many bird in one run... 

And since you refuse adding a flag to do it.. you force us, and I do believe most of us, to just use a [plugin](https://github.com/gajus/write-file-webpack-plugin) even if just to save us few seconds every time.. cause these few second pile up eventually, and personally I hate wasting my time for no god reason.

**Comment by chrisisler** [comment]
Just wanted to chime in to say again what has already been said: Django + webpack --watch is annoying without the awesomeness of hot reloading. Haven't tried @gajus plugin. This should be a core feature, or at least mentioned somewhere in the webpack-dev-server documentation.

**Comment by trusktr** [comment]
> But why would you need to access the generated assets?

Because I'm developing a library which I need to test in a browser, but I also need to import it from the file system in Node.js applications (f.e. other Electron apps that I'm also testing).

**Comment by trusktr** [comment]
Essentially `webpack-dev-server` is designed to be used while developing apps, not while developing libraries.

**Comment by damianobarbati** [comment]
Facing same issue. @sokra @trusktr in my scenario I'm trying to write a library and in the meantime watch it in an app:

```
const pkg = require('./package.json');
const webpack = require('webpack');
const path = require('path');

module.exports = [{
    entry: './src/index.js',
    output: {
        path: path.resolve(__dirname, './dist'),
        filename: 'dist.min.js',
        library: pkg.name,
        libraryTarget: 'umd'
    },
    watch: true,
    module: {
        rules: [{
            test: /\.(js|jsx)$/,
            exclude: /node_modules/,
            use: [{
                loader: 'babel-loader',
                options: {
                    cacheDirectory: true,
                },
            }],
        }],
    },
    externals: {
    },
},{
    entry: './demo/index.js',
    output: {
        publicPath: '/',
        path: __dirname + '/demo',
        filename: 'bundle.js'
    },
    module: {
        rules: [{
            test: /\.(js|jsx)$/,
            exclude: /node_modules/,
            use: [{
                loader: 'babel-loader',
                options: {
                    cacheDirectory: true,
                },
            }],
        }],
    },
    resolve: {
        alias: {
            Form: path.resolve(__dirname, './dist/dist.min.js'),
        },
    },
    plugins: [
        new webpack.HotModuleReplacementPlugin(),
    ],
    devServer: {
        contentBase: './demo',
        hot: true
    },
}];
```

The latter does not find the "resolved" => "Form" bundle on disk because it is not written.
If I trigger separated the build, subsequent changes to form bundle do not trigger dev-server "watch".

:(

**Comment by smac89** [comment]
```js
devServer: { // https://webpack.js.org/configuration/dev-server
    writeToDisk: true,
},
```

**Comment by IronSean** [comment]
In my case, I don't export an `index.html` to my `/dist` folder, I use `HtmlWebpackPlugin` to output template `.cshtml` files that are served by a .Net app. The `bundle.js` files are referenced there, and I'm having trouble figuring out how to get `webpack-dev-server` to output them. `writeToDisk` doesn't output them, and even `localhost:8080/webpack-dev-server` doesn't list them, as if they're not even being generated by webpack-dev-server at all.

**Comment by Mr-Lion** [comment]
> In webpack config:
> 
> ```js
> {
>    ...
>     devServer: { // https://webpack.js.org/configuration/dev-server/#devserverwritetodisk-
>         writeToDisk: true,
>     },
> }
> ```

Hello friends,

I've been dealing with the problem for several weeks and look in vain for a sensible solution.

The one with the "writeToDisk: true" definitely doesn't work.
No matter what I did and how I tried every possible start, I could not write the "css" for the style on the disc.

Furthermore, the webpack-dev-server does not work at all with a multi-config. Here comes the mistake:

![image](https://user-images.githubusercontent.com/6929357/144687323-0297009c-90af-4ba4-9f27-971d34d4828f.png)

I don't understand why the "CSS" files are written into the memory. The developer must have really burned his brain. If you search the internet for this problem, you will find millions of complaints where people are looking for a solution. One should make the decision easy for a user.

I urgently ask for help!
I thank you in advance.
